internal/rbd/sms_controllerserver.go

```diff
@@ -0,0 +1,220 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// defaultMaxResults is the internal limit for max results in metadata streaming
+	// This is set to a reasonable value to prevent excessive memory usage
+	// and ensure efficient processing of metadata requests.
+	// TODO: Consider making this configurable via a cmdline flag.
+	defaultMaxResults int32 = 128
+)
+
+// normalizeMaxResults normalizes the maxResults parameter. It handles zero values,
+// applies internal limits, and provides logging for debugging.
+func normalizeMaxResults(ctx context.Context, requestedMaxResults int32) int32 {
+	// Handle zero case - user wants default behavior
+	if requestedMaxResults == 0 {
+		log.DebugLog(ctx, "maxResults was 0, using internal default: %d", defaultMaxResults)
+
+		return defaultMaxResults
+	}
+
+	// Apply internal max limit for memory management
+	if requestedMaxResults > defaultMaxResults {
+		log.DebugLog(ctx, "maxResults limited from %d to %d (internal max limit)",
+			requestedMaxResults, defaultMaxResults)
+
+		return defaultMaxResults
+	}
+	log.DebugLog(ctx, "using requested maxResults: %d", requestedMaxResults)
+
+	return requestedMaxResults
+}
+
+func validateMetadataAllocatedReq(req *csi.GetMetadataAllocatedRequest) error {
+	if req.GetSnapshotId() == "" {
+		return status.Error(codes.InvalidArgument, "snapshot ID cannot be empty")
+	}
+
+	if req.GetSecrets() == nil || len(req.GetSecrets()) == 0 {
+		return status.Error(codes.InvalidArgument, "secrets cannot be nil or empty")
+	}
+
+	if req.GetMaxResults() < 0 {
+		return status.Error(codes.InvalidArgument, "max results cannot be negative")
+	}
+
+	if req.GetStartingOffset() < 0 {
+		return status.Error(codes.OutOfRange, "starting offset cannot be negative")
+	}
+
+	return nil
+}
+
+// GetMetadataAllocated streams the allocated block metadata for a snapshot.
+func (cs *ControllerServer) GetMetadataAllocated(
+	req *csi.GetMetadataAllocatedRequest,
+	stream csi.SnapshotMetadata_GetMetadataAllocatedServer,
+) error {
+	if err := validateMetadataAllocatedReq(req); err != nil {
+		return err
+	}
+
+	snapshotID := req.GetSnapshotId()
+
+	ctx := context.WithValue(stream.Context(), log.ReqID, req.GetSnapshotId())
```

Comment:
log.ReqID will not be set for this one unless its added to middle ware? and also set the SnapshotID in the middleware itself?

Reply:
> log.ReqID will not be set for this one unless its added to middle ware? and also set the SnapshotID in the middleware itself?

Yes, its not logged :disappointed: ,

we need a new middle ware for streamserver interceptor, current one does not work these stream rpcs.
I'll send a separate pr for that.